content/posts/deploying-application-on-k8s-with-helm/index.md

```diff
@@ -0,0 +1,431 @@
+
+We need a solution to manage all of them in a common template => Helm can solve this problem
+
+Before go to next part, you need to download and install minikube on your local device.
+
+## Introduction to helm chart
+**Helm is a tool that streamlines installing and managing Kubernetes applications.** To work with helm chart, we need to install helm command. This link is about how to install: <a>https://helm.sh/docs/intro/install</a>
+
+I will introduce you to the structure of the helm chart. To create a chart, we can use my example command:
+```bash
+
+  helm create mynginx
+
+```
+That command will create an mynginx folder. Let take a look in mynginx folder, we have some folders and files:
+- charts: this folder will contain another subchart which our chart will depend on. I will talk more about subchart it in the next section
+- templates: this folder contains yaml files about the k8s resource. We can add more, remove or modify the files in this folder
+- Chart.yaml: this file is the description about the chart information
+- _helpers.tpl: this file defines some yaml key-value which we can import to yaml files in templates folders
+
+Let take a look about the file template/deployments.yaml:
+
+```yaml
+
+  apiVersion: apps/v1
+  kind: Deployment
+  metadata:
+    name: {{ include "mynginx.fullname" . }}
+    labels:
+      {{- include "mynginx.labels" . | nindent 4 }}
+  spec:
+    {{- if not .Values.autoscaling.enabled }}
+    replicas: {{ .Values.replicaCount }}
+    {{- end }}
+    selector:
+      matchLabels:
+        {{- include "mynginx.selectorLabels" . | nindent 6 }}
+    template:
+      metadata:
+        {{- with .Values.podAnnotations }}
+        annotations:
+          {{- toYaml . | nindent 8 }}
+        {{- end }}
+        labels:
+          {{- include "mynginx.selectorLabels" . | nindent 8 }}
+      spec:
+        {{- with .Values.imagePullSecrets }}
+        imagePullSecrets:
+          {{- toYaml . | nindent 8 }}
+        {{- end }}
+        serviceAccountName: {{ include "mynginx.serviceAccountName" . }}
+        securityContext:
+          {{- toYaml .Values.podSecurityContext | nindent 8 }}
+        containers:
+          - name: {{ .Chart.Name }}
+            securityContext:
+              {{- toYaml .Values.securityContext | nindent 12 }}
+            image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
+            imagePullPolicy: {{ .Values.image.pullPolicy }}
+  ..........................................................
+
+```
+
+This yaml file looks like the template of k8s resource. But it has some {{ }}, the statement in {{ }} can be a value, operator, expression,.... Let me explain some popular thing:
```

Comment:
This yaml file looks like the template of k8s resource, except we see alot of this `{{ }}` syntax. The statement in {{ }} can be a value, operator, expression, etc. Let me explain some popular uses: